src/tir/transforms/local_pad.cc

```diff
@@ -0,0 +1,302 @@
+    CheckInitValue_<FloatImmNode>(op->value);
+    return StmtVisitor::VisitStmt_(op);
+  }
+  template <typename ImmNodeType>
+  void CheckInitValue_(const PrimExpr& rhs) {
+    if (const ImmNodeType* const rhs_val = rhs.as<ImmNodeType>()) {
+      if (init_constexpr_.defined()) {
+        if (const ImmNodeType* const init_val = init_constexpr_.as<ImmNodeType>()) {
+          if (rhs_val->value != init_val->value) {
+            init_with_single_constexpr_ = false;
+          }
+        } else {
+          init_with_single_constexpr_ = false;
+        }
+      } else {
+        init_with_single_constexpr_ = true;
+        init_constexpr_ = rhs;
+      }
+    }
+  }
+  void operator()(const Stmt& stmt) {
+    StmtVisitor::operator()(stmt);
+    if (!init_with_single_constexpr_) {
+      init_constexpr_ = PrimExpr();
+    }
+  }
+
+  bool init_with_single_constexpr_ = false;
+  PrimExpr init_constexpr_;
+
+  friend class LocalPadder;
+};
+
+/*!
+ * \brief Split a predicate into inlinable and non-inlinable component.
+ *
+ *        We refer to "inlinable predicate" as
+ *
+ *            if (predicate) A = ...;
+ *            ↓
+ *            A = predicate ? ... : init_constexpr;
+ *
+ *        Note that not all predicates can be inlined. For example, if a predicate is there to guard
+ *        against out-of-boundary accesses to local/shared variables, then it cannot be inlined.
+ */
+class PredicateInliner : public StmtExprVisitor {
+ private:
+  explicit PredicateInliner(const Stmt& body_stmt) : body_stmt_(body_stmt) {}
+
+#define VISIT_PREDICATE(OpType)                      \
+  void VisitExpr_(const OpType##Node* op) final {    \
+    OpType predicate = GetRef<OpType>(op);           \
+    if (CanInlinePredicate_<OpType##Node>(op)) {     \
+      inlinable_predicates_.push_back(predicate);    \
+    } else {                                         \
+      non_inlinable_residuals_.push_back(predicate); \
+    }                                                \
+  }
+  VISIT_PREDICATE(LT)
+  VISIT_PREDICATE(LE)
+  VISIT_PREDICATE(GT)
+  VISIT_PREDICATE(GE)
+#undef VISIT_PREDICATE
+
+  void VisitStmt_(const BufferStoreNode* op) final {
+    if (op->indices.size() != 1) {
+      return StmtVisitor::VisitStmt_(op);
+    }
+    CHECK(op->buffer.scope() == "shared" || op->buffer.scope() == "local");
+    if (StructuralEqual()(op->indices[0], predicate_lhs_)) {
+      predicate_inlinable_ = false;
+    }
+    return StmtVisitor::VisitStmt_(op);
+  }
+  /*!
+   * \brief Check if a predicate can be inlined.
+   */
+  template <typename OpNodeType>
+  bool CanInlinePredicate_(const OpNodeType* op) {
+    predicate_inlinable_ = true;
+    predicate_lhs_ = op->a;
+    VisitStmt(body_stmt_);
+    return predicate_inlinable_;
+  }
+
+  Stmt body_stmt_;
+  std::vector<PrimExpr> inlinable_predicates_, non_inlinable_residuals_;
+  bool predicate_inlinable_;
+  PrimExpr predicate_lhs_;
+
+  friend class LocalPadder;
+};
```

Comment:
Generally, I would love to propose that we restructure the logic of this class a little bit.

Looks like the analyzer is interested in the following patterns:

```python
if A </<=/>=/> X:
  B[Y] = ...
```

If so, there isn't much reason to use a visitor pattern because recursion didn't actually happen. Instead, let's go with a more plain and readable fashion, for example

```C++
// inputs:
IfThenElse if_then_else;
// extract the lhs & rhs of the if-condition
PrimExpr predicate_lhs{nullptr};
PrimExpr predicate_rhs{nullptr};
if (const auto *op = if_then_else->condition.as<LENode>()) {
  predicate_lhs = op->a;
  predicate_rhs = op->a;
} else if (...) {
  // use a macro or something to deal with LT, GE, GT
}
// then let's analyze the body statement
const BufferStoreNode* buffer_store = if_then_else->then_case.as<BufferStoreNode>();
ICHECK(buffer_store);
if (StructuralEqual()(buffer_store->indices[0], predicate_lhs)) {
  ... // some logic here
} else {
  ... // some logic here
}
```

Reply:
I am afraid we cannot directly do it like this. The reason is because the predicates are usually combined together into a single one and hence we need some way of splitting them. The implementation you provide might not be able to handle situations like the following:

```C++
if (inlineable_predicate1 && non_inlineable_predicate2 && inlineable_predicate3)
  A_shared[...] = A[...];

|  // after transformation

if (noninlineable_predicate2
  A_shared[...] = inlineable_predicate1 && inlineable_predicate3 ? A[...] : padded_value;
```